<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

## Connect to Hyperliquid API 
### setup function

In [0]:
#| echo: false
#| output: asis
show_doc(setup)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L31){target="_blank" style="float:right; font-size:smaller"}

### setup

```python
def setup(
    base_url:NoneType=None, skip_ws:bool=False, perp_dexs:NoneType=None, config:str='../config_hyperliquid.json',
    debug:bool=False
):
```

*Call self as a function.*

### Example

In [7]:
#| eval: false
address, info, exchange = setup(base_url=constants.MAINNET_API_URL, skip_ws=True)

## HIP-3 / RWA dex support
### Helper functions for resolving builder-dex tickers (e.g. XLE → xyz:XLE)

Hyperliquid hosts third-party perpetual markets on "builder dexes" (HIP-3).
These tickers use a `dex:COIN` naming convention (e.g. `xyz:XLE`).
The helpers below let every data-download function transparently resolve
a plain ticker like `"XLE"` to `"xyz:XLE"` and fall back to a raw HTTP
request when the Python SDK's internal name map doesn't know the coin.

In [0]:
#| echo: false
#| output: asis
show_doc(resolve_hyperliquid_ticker)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L78){target="_blank" style="float:right; font-size:smaller"}

### resolve_hyperliquid_ticker

```python
def resolve_hyperliquid_ticker(
    coin, info:NoneType=None, base_url:NoneType=None
):
```

*Resolve a plain ticker to its dex-prefixed form if needed.*

- If *coin* already contains ':' it is returned as-is.
- If *coin* exists in the native perp universe it is returned as-is.
- Otherwise each HIP-3 dex universe is searched for ``{dex}:{coin}``.
- Results are cached so repeated calls are cheap.

Args:
    coin (str): Ticker such as ``"ETH"`` or ``"XLE"``.
    info (Info, optional): SDK Info client (used for native universe check).
    base_url (str, optional): API base URL override.

Returns:
    str: Resolved ticker (e.g. ``"xyz:XLE"``).  Falls back to *coin*
         unchanged if nothing is found.

In [0]:
#| echo: false
#| output: asis
show_doc(hyperliquid_perp_dexs)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L73){target="_blank" style="float:right; font-size:smaller"}

### hyperliquid_perp_dexs

```python
def hyperliquid_perp_dexs(
    base_url:NoneType=None
):
```

*Return list of perp dex name strings ('' = native, 'xyz', etc.).*

## Get historical prerpetual price data
### retrive_hyperliquid_perp_price function

In [0]:
#| echo: false
#| output: asis
show_doc(retrieve_hyperliquid_perp_price)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L138){target="_blank" style="float:right; font-size:smaller"}

### retrieve_hyperliquid_perp_price

```python
def retrieve_hyperliquid_perp_price(
    coin:str='ETH', interval:str='1h', end_date:str='2026-07-26T23:42:10Z', start_date:str='2026-07-24T23:42:10Z',
    info:NoneType=None
):
```

*Retrieves historical candle data from Hyperliquid for a given coin and time interval.*

Args:
    coin (str, optional): Coin symbol (e.g. "ETH", "XLE"). Defaults to "ETH".
        Plain RWA tickers are auto-resolved (e.g. "XLE" → "xyz:XLE").
    interval (str, optional): Candle interval ("1m", "5m", "15m", "1h", "4h", "1d"). Defaults to "1h".
    end_date (str, optional): End datetime in ISO 8601 format. Defaults to current UTC time.
    start_date (str, optional): Start datetime in ISO 8601 format. Defaults to 2 days before end_date.
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    pandas.DataFrame: DataFrame containing the OHLCV data with columns:
        - datetime: Timestamp for the candle (UTC)
        - open: Opening price of the interval
        - high: Highest traded price in the interval
        - low: Lowest traded price in the interval
        - close: Closing price of the interval
        - volume: Trading volume in the interval
        - coin: Coin symbol
    Returns None if the API request fails or returns no data.

Notes:
    - All datetime values are in UTC timezone
    - Requires Hyperliquid Info client to be initialized

### Example

You will need a api key and secret key from the Hyperliquid API. Store this into the `config_hyperliquid.json` file in the same directory as your script. See: https://app.hyperliquid.xyz/API

The best is to first call the [`setup`](https://silvaac.github.io/token_data/hyperliquid.html#setup) function once to initialize the Hyperliquid Info client. Then pass "info" to any function that requires it. That will avoid calling the [`setup`](https://silvaac.github.io/token_data/hyperliquid.html#setup) function multiple times.

Typical usage:

In [10]:
#| eval: false
perp = retrieve_hyperliquid_perp_price(coin="ETH", interval="1h",info=info)
print(perp.head())

             datetime    open    high     low   close     volume coin
0 2026-07-24 23:00:00  1858.8  1861.1  1858.1  1861.0  2792.8168  ETH
1 2026-07-25 00:00:00  1860.7  1863.7  1856.3  1859.3  4137.4891  ETH
2 2026-07-25 01:00:00  1859.3  1862.6  1857.4  1859.3  4557.0876  ETH
3 2026-07-25 02:00:00  1859.3  1860.7  1857.3  1858.3  4404.2636  ETH
4 2026-07-25 03:00:00  1858.3  1859.5  1854.8  1858.3  4298.4131  ETH


## List of spot tickers
### spot_tickers function

In [0]:
#| echo: false
#| output: asis
show_doc(spot_tickers)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L237){target="_blank" style="float:right; font-size:smaller"}

### spot_tickers

```python
def spot_tickers(
    coin:str='ETH', base:str='USDC', info:NoneType=None
):
```

*Retrieves current tickers for a given coin.*

Args:
    coin (str, optional): Coin symbol (e.g. "ETH"). Defaults to "ETH".
    base (str, optional): Coin symbol (e.g. "USDC"). Defaults to "USDC".
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    spot ticker (non intuitive symbol)
    Returns None if the API request fails or returns no data.

Notes:
    - Requires Hyperliquid Info client to be initialized

### Example

In [12]:
#| eval: false
stk = spot_tickers(info=info,coin="ETH",base="USDC")
print(f'spot ticker for ETH: ', stk)

spot ticker for ETH:  @151


## Get historical spot price
### retrieve_hyperliquid_spot_price function

In [0]:
#| echo: false
#| output: asis
show_doc(retrieve_hyperliquid_spot_price)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L294){target="_blank" style="float:right; font-size:smaller"}

### retrieve_hyperliquid_spot_price

```python
def retrieve_hyperliquid_spot_price(
    coin:str='ETH', base:str='USDC', interval:str='1h', end_date:str='2026-07-26T23:42:10Z',
    start_date:str='2026-07-24T23:42:10Z', info:NoneType=None, recheck:bool=False
):
```

*Call self as a function.*

### Example

The ticker for Ethereum (ETH) is "@151" ("UETH"). There are other unusual choices but the function handles these choices by changing the ticker inside the function. 

In [14]:
#| eval: false
spot=retrieve_hyperliquid_spot_price(info=info)
print(spot.tail())

              datetime    open    high     low   close    volume coin
44 2026-07-26 19:00:00  1913.0  1913.8  1910.2  1912.5  336.0437  ETH
45 2026-07-26 20:00:00  1912.5  1912.9  1910.2  1912.1  136.4614  ETH
46 2026-07-26 21:00:00  1912.8  1927.2  1912.2  1923.9  139.3486  ETH
47 2026-07-26 22:00:00  1924.4  1951.0  1924.4  1948.5  461.1441  ETH
48 2026-07-26 23:00:00  1951.1  1964.7  1948.5  1959.5  142.2609  ETH


If you want to find out which ticker is used for a given coin, you can use the [`spot_tickers`](https://silvaac.github.io/token_data/hyperliquid.html#spot_tickers) function. For example, for Ethereum (ETH) quoted in USDC is:

In [15]:
#| eval: false
stk = spot_tickers(info=info,coin="ETH",base="USDC")
print(f'spot ticker for ETH: ', stk)

spot ticker for ETH:  @151


## List all tokens in Hyperliquid
### hyperliquid_tokens function

In [0]:
#| echo: false
#| output: asis
show_doc(hyperliquid_tokens)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L330){target="_blank" style="float:right; font-size:smaller"}

### hyperliquid_tokens

```python
def hyperliquid_tokens(
    info:NoneType=None, rm_delisted:bool=True, dex:NoneType=None
):
```

*List perpetual tokens available on Hyperliquid.*

Args:
    info (Info, optional): SDK Info client. Created automatically if *None*.
    rm_delisted (bool): Remove delisted and isolated-only tokens (default *True*).
    dex (str, optional): Which perp dex universe to query.

        - ``None`` (default) – native perps only (unchanged legacy behaviour).
        - A dex name string (e.g. ``"xyz"``) – that builder-dex universe.
          Returned ``name`` values are already prefixed (e.g. ``"xyz:XLE"``).
        - ``"all"`` – concatenates native **and** every HIP-3 dex universe.

Returns:
    pandas.DataFrame with columns ``szDecimals``, ``name``, ``maxLeverage``, etc.

### Example

In [17]:
#| eval: false
tokens = hyperliquid_tokens(info)
print(tokens)

     szDecimals   name  maxLeverage  marginTableId  isDelisted  onlyIsolated  \
0             5    BTC           40             56       False         False   
1             4    ETH           25             55       False         False   
2             2   ATOM            5              5       False         False   
4             1   DYDX            5              5       False         False   
5             2    SOL           20             54       False         False   
..          ...    ...          ...            ...         ...           ...   
226           2   DASH            5              5       False         False   
227           0    SKR            3              3       False         False   
228           0  AZTEC            3              3       False         False   
229           0   CHIP            3              3       False         False   
230           0   GRAM            5              5       False         False   

    marginMode  
0          NaN  
1    

## Get funding rate history
### retrieve_hyperliquid_funding_history function

In [0]:
#| echo: false
#| output: asis
show_doc(funding_calc)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L390){target="_blank" style="float:right; font-size:smaller"}

### funding_calc

```python
def funding_calc(
    rate, premium, max_rate:float=0.0005, min_rate:float=-0.0005
):
```

*Call self as a function.*

In [0]:
#| echo: false
#| output: asis
show_doc(retrieve_hyperliquid_funding_history)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L394){target="_blank" style="float:right; font-size:smaller"}

### retrieve_hyperliquid_funding_history

```python
def retrieve_hyperliquid_funding_history(
    coin:str='ETH', end_date:str='2026-07-26T23:42:10Z', start_date:str='2026-07-24T23:42:10Z', info:NoneType=None,
    calc:bool=False
):
```

*Retrieves funding rate history from Hyperliquid for a given coin and time period.*

Args:
    coin (str, optional): Coin symbol (e.g. "ETH", "XLE"). Defaults to "ETH".
        Plain RWA tickers are auto-resolved (e.g. "XLE" → "xyz:XLE").
    end_date (str, optional): End datetime in ISO 8601 format. Defaults to current UTC time.
    start_date (str, optional): Start datetime in ISO 8601 format. Defaults to 7 days before end_date.
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    pandas.DataFrame: DataFrame containing the funding history with columns:
        - datetime: Timestamp for the funding rate (UTC)
        - funding_rate: The funding rate value
        - premium: The premium component
        - coin: Coin symbol
    Returns None if the API request fails or returns no data.

Notes:
    - All datetime values are in UTC timezone
    - Funding rates are typically updated every hour
    - Requires Hyperliquid Info client to be initialized

### Example

In [20]:
#| eval: false
f_r = retrieve_hyperliquid_funding_history(info=info,calc=True)
print(f_r.tail())

                  datetime  funding_rate   premium coin  fund_calc
43 2026-07-26 19:00:00.031      0.000013 -0.000234  ETH   0.000013
44 2026-07-26 20:00:00.057      0.000013 -0.000285  ETH   0.000012
45 2026-07-26 21:00:00.001      0.000013 -0.000335  ETH   0.000012
46 2026-07-26 22:00:00.023      0.000013 -0.000301  ETH   0.000012
47 2026-07-26 23:00:00.005      0.000013 -0.000202  ETH   0.000013


## Unified function for easy data retrieval
### retrieve_hyperliquid_data function

In [0]:
#| echo: false
#| output: asis
show_doc(retrieve_hyperliquid_data)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L498){target="_blank" style="float:right; font-size:smaller"}

### retrieve_hyperliquid_data

```python
def retrieve_hyperliquid_data(
    ticker:str='ETH', data_type:str='perp', start_date:NoneType=None, end_date:NoneType=None, lookback:int=2,
    interval:str='1h', base:str='USDC', round_to_hour:bool=False, info:NoneType=None
):
```

*Unified function to retrieve funding rates, perpetual prices, or spot prices from Hyperliquid.*

Args:
    ticker (str, optional): Coin symbol (e.g. "ETH", "BTC"). Defaults to "ETH".
    data_type (str, optional): Type of data to retrieve - "funding", "perp", or "spot". Defaults to "perp".
    start_date (str, optional): Start date as string. Can be:
        - ISO format: "2024-01-15T10:30:00Z" or "2024-01-15T10:30:00"
        - Date only: "2024-01-15"
        - If None, calculated from lookback. Defaults to None.
    end_date (str, optional): End date as string (same formats as start_date).
        - If None, uses current UTC time. Defaults to None.
    lookback (int, optional): Number of days to look back from end_date if start_date is None. 
        Defaults to 2.
    interval (str, optional): Candle interval for perp/spot data ("1m", "5m", "15m", "1h", "4h", "1d"). 
        Defaults to "1h". Not used for funding rates.
    base (str, optional): Base currency for spot prices (e.g. "USDC"). Defaults to "USDC".
        Not used for perp or funding rates.
    round_to_hour (bool, optional): If True, rounds start_date and end_date to nearest hour.
        Useful for funding rates which update hourly. Defaults to False.
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    pandas.DataFrame: DataFrame containing the requested data with appropriate columns:
        - For "funding": datetime, funding_rate, premium, coin
        - For "perp": datetime, open, high, low, close, volume, coin
        - For "spot": datetime, open, high, low, close, volume, coin
    Returns None if the API request fails or returns no data.

Examples:
    # Get 7 days of funding rates for ETH, rounded to hour
    df = retrieve_hyperliquid_data("ETH", "funding", lookback=7, round_to_hour=True, info=info)

    # Get perp prices between specific dates with 4h interval
    df = retrieve_hyperliquid_data("BTC", "perp", 
                                  start_date="2024-01-01", 
                                  end_date="2024-01-15",
                                  interval="4h", info=info)

    # Get spot prices for last 30 days with 1h interval
    df = retrieve_hyperliquid_data("ETH", "spot", lookback=30, 
                                  interval="1h", base="USDC", info=info)

Notes:
    - All datetime values are in UTC timezone
    - Valid data_type values: "funding", "perp", "spot"
    - Funding rates are updated hourly, so round_to_hour=True is recommended
    - Requires Hyperliquid Info client to be initialized

### Example

In [22]:
#| eval: false
## Example 1: Get funding rates for last 7 days, rounded to hour
funding_df = retrieve_hyperliquid_data(
    ticker="ETH",
    data_type="funding",
    lookback=7,
    round_to_hour=True,
    info=info
)
print(funding_df.tail())

               datetime  funding_rate   premium coin  fund_calc
163 2026-07-26 19:00:00      0.000013 -0.000234  ETH   0.000013
164 2026-07-26 20:00:00      0.000013 -0.000285  ETH   0.000013
165 2026-07-26 21:00:00      0.000013 -0.000335  ETH   0.000013
166 2026-07-26 22:00:00      0.000013 -0.000301  ETH   0.000013
167 2026-07-26 23:00:00      0.000013 -0.000202  ETH   0.000013


In [23]:
#| eval: false
## Example 2: Get perpetual prices with specific dates and 4h interval
perp_df = retrieve_hyperliquid_data(
    ticker="BTC",
    data_type="perp",
    start_date="2026-01-01",
    end_date="2026-01-15",
    interval="4h",
    info=info
)
print(perp_df.tail())

              datetime     open     high      low    close       volume coin
80 2026-01-14 08:00:00  95182.0  95302.0  94714.0  94773.0   2307.95490  BTC
81 2026-01-14 12:00:00  94773.0  97102.0  94647.0  96737.0  13517.53853  BTC
82 2026-01-14 16:00:00  96737.0  97765.0  96263.0  97241.0  11110.15613  BTC
83 2026-01-14 20:00:00  97241.0  97949.0  96799.0  96917.0   6714.43893  BTC
84 2026-01-15 00:00:00  96917.0  96975.0  95747.0  95970.0   5284.50121  BTC


In [24]:
#| eval: false
## Example 3: Get spot prices for last 30 days
spot_df = retrieve_hyperliquid_data(
    ticker="ETH",
    data_type="spot",
    lookback=30,
    interval="1h",
    base="USDC",
    info=info
)
print(spot_df.tail())

               datetime    open    high     low   close    volume coin
716 2026-07-26 19:00:00  1913.0  1913.8  1910.2  1912.5  336.0437  ETH
717 2026-07-26 20:00:00  1912.5  1912.9  1910.2  1912.1  136.4614  ETH
718 2026-07-26 21:00:00  1912.8  1927.2  1912.2  1923.9  139.3486  ETH
719 2026-07-26 22:00:00  1924.4  1951.0  1924.4  1948.5  461.1441  ETH
720 2026-07-26 23:00:00  1951.1  1964.7  1948.5  1959.5  142.2609  ETH


In [25]:
#| eval: false
## Example 4: Using datetime strings with time information
data_df = retrieve_hyperliquid_data(
    ticker="ETH",
    data_type="perp",
    lookback=2,
    interval="1h",
    info=info
)
print(data_df.tail())

              datetime    open    high     low   close      volume coin
44 2026-07-26 19:00:00  1914.0  1915.4  1911.5  1913.6   2512.5447  ETH
45 2026-07-26 20:00:00  1913.6  1914.0  1910.9  1912.8   2440.3697  ETH
46 2026-07-26 21:00:00  1912.7  1929.9  1912.7  1925.3  13779.9050  ETH
47 2026-07-26 22:00:00  1925.3  1951.1  1925.3  1949.0  47844.9203  ETH
48 2026-07-26 23:00:00  1949.0  1967.9  1949.0  1960.5  20710.9213  ETH


In [26]:
#| eval: false
## Example 5: Get funding rates with date-only strings
funding_df = retrieve_hyperliquid_data(
    ticker="BTC",
    data_type="funding",
    start_date="2025-08-29",
    end_date="2025-09-01",
    round_to_hour=True,
    info=info
)
print(funding_df)

              datetime  funding_rate   premium coin  fund_calc
0  2025-08-29 00:00:00      0.000013  0.000251  BTC   0.000013
1  2025-08-29 01:00:00      0.000013  0.000234  BTC   0.000013
2  2025-08-29 02:00:00      0.000013  0.000124  BTC   0.000013
3  2025-08-29 03:00:00      0.000013  0.000068  BTC   0.000013
4  2025-08-29 04:00:00      0.000013  0.000086  BTC   0.000013
..                 ...           ...       ...  ...        ...
67 2025-08-31 19:00:00      0.000013 -0.000035  BTC   0.000013
68 2025-08-31 20:00:00      0.000013 -0.000082  BTC   0.000013
69 2025-08-31 21:00:00      0.000013 -0.000028  BTC   0.000013
70 2025-08-31 22:00:00      0.000013 -0.000059  BTC   0.000013
71 2025-08-31 23:00:00      0.000013 -0.000012  BTC   0.000013

[72 rows x 5 columns]


In [27]:
#| eval: false
data_df = retrieve_hyperliquid_data(
    ticker="ETH",
    data_type="perp",
    start_date="2026-06-15T10:30:00",
    end_date="2026-06-20T15:45:00",
    interval="1h",
    info=info
)
data_df.tail()

,datetime,open,high,low,close,volume,coin
121,2026-06-20 11:00:00,1727.5,1731.1,1725.9,1727.1,2295.1586,ETH
122,2026-06-20 12:00:00,1727.1,1730.5,1723.6,1726.7,20751.8753,ETH
123,2026-06-20 13:00:00,1726.6,1726.6,1712.7,1717.3,18955.0975,ETH
124,2026-06-20 14:00:00,1717.4,1745.5,1708.0,1733.1,21337.8294,ETH
125,2026-06-20 15:00:00,1733.2,1750.0,1729.3,1739.6,46985.6872,ETH


## Order book data
### retrieve_hyperliquid_l2_snapshot

The will return the snapshot of the order book for the specified ticker in the moment you call the function. It does not provide historical data. 

This is good only for perpetual markets.

In [28]:
def unix_to_datetime(t):
    return datetime.fromtimestamp(t/1000).strftime("%Y-%m-%d %H:%M:%S.%f")
#The funding rates are reset every hour.
#t = datetime.datetime.fromtimestamp(i['time']/1000).strftime("%Y-%m-%d %H:%M:%S.%f")
#print(t)

In [0]:
#| echo: false
#| output: asis
show_doc(retrieve_hyperliquid_l2_snapshot)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L628){target="_blank" style="float:right; font-size:smaller"}

### retrieve_hyperliquid_l2_snapshot

```python
def retrieve_hyperliquid_l2_snapshot(
    coin:str='ETH', info:NoneType=None
):
```

*Retrieves current L2 order book snapshot from Hyperliquid for a given coin.*

Args:
    coin (str, optional): Coin symbol (e.g. "ETH", "BTC", "XLE"). Defaults to "ETH".
        Plain RWA tickers are auto-resolved (e.g. "XLE" → "xyz:XLE").
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    pandas.DataFrame: DataFrame containing the order book snapshot with columns:
        - datetime: Timestamp of the snapshot (UTC)
        - side: Order side ("bid" or "ask")
        - price: Price level
        - size: Total size at this price level
        - num_orders: Number of orders at this price level
    Returns None if the API request fails or returns no data.

Notes:
    - This is a snapshot at the moment the function is called
    - All datetime values are in UTC timezone
    - Bids are sorted from highest to lowest price
    - Asks are sorted from lowest to highest price
    - Requires Hyperliquid Info client to be initialized

### Example

In [30]:
#| eval: false
# Example usage: Get L2 order book snapshot for ETH
l2_snapshot = retrieve_hyperliquid_l2_snapshot(coin="ETH", info=info)
print(l2_snapshot.head(5))
print(l2_snapshot.tail(5))

# Check the structure
print(f"\nTotal levels: {len(l2_snapshot)}")
print(f"Bids: {len(l2_snapshot[l2_snapshot['side'] == 'bid'])}")
print(f"Asks: {len(l2_snapshot[l2_snapshot['side'] == 'ask'])}")
print(f"Snapshot time: {l2_snapshot['datetime'].iloc[0]}")

                 datetime side   price     size  num_orders
0 2026-07-26 23:16:41.579  bid  1960.5   4.3721           2
1 2026-07-26 23:16:41.579  bid  1960.4  34.3144           4
2 2026-07-26 23:16:41.579  bid  1960.3   1.4367           1
3 2026-07-26 23:16:41.579  bid  1960.2  16.8822           4
4 2026-07-26 23:16:41.579  bid  1960.1  32.5682          10
                  datetime side   price      size  num_orders
35 2026-07-26 23:16:41.579  ask  1962.1  202.2812          11
36 2026-07-26 23:16:41.579  ask  1962.2  224.6602          15
37 2026-07-26 23:16:41.579  ask  1962.3  231.3283          11
38 2026-07-26 23:16:41.579  ask  1962.4  198.4616          13
39 2026-07-26 23:16:41.579  ask  1962.5  265.0086          15

Total levels: 40
Bids: 20
Asks: 20
Snapshot time: 2026-07-26 23:16:41.579000


## Mid prices
### hyperliquid_mids

In [0]:
#| echo: false
#| output: asis
show_doc(hyperliquid_mids)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L715){target="_blank" style="float:right; font-size:smaller"}

### hyperliquid_mids

```python
def hyperliquid_mids(
    coin:NoneType=None, info:NoneType=None, typecast_to_float:bool=True
):
```

*Retrieves current mid prices from Hyperliquid for specified coins or all available coins.*

Args:
    coin (str, optional): Coin symbol (e.g. "ETH", "BTC", "XLE").
        Plain RWA tickers are auto-resolved (e.g. "XLE" → "xyz:XLE").
        For HIP-3/RWA coins not in the native allMids endpoint the mid
        is computed from the L2 book.
        If None, retrieves mids for all available coins. Defaults to None.
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    dictionary: Dictionary containing mid prices with keys as coin symbols and values as mid prices
    Returns None if the API request fails or returns no data.

Examples:
    # Get mid prices for specific coins
    dic = hyperliquid_mids(info=info)

Notes:
    - This is a snapshot at the moment the function is called
    - Mid price is calculated as (best_bid + best_ask) / 2
    - Requires Hyperliquid Info client to be initialized

### Example

In [32]:
#| eval: false
hyperliquid_mids(coin="ETH",info=info)

1960.55

In [33]:
#| eval: false
hyperliquid_mids(info=info)

{'#5090': 0.5,
 '#5091': 0.5,
 '#5100': 0.783925,
 '#5101': 0.216075,
 '#5110': 0.00553,
 '#5111': 0.99447,
 '#5120': 0.184345,
 '#5121': 0.815655,
 '#9290': 0.980635,
 '#9291': 0.019365,
 '#9300': 0.982445,
 '#9301': 0.017555,
 '#9310': 0.986455,
 '#9311': 0.013545,
 '#9320': 0.894995,
 '#9321': 0.105005,
 '#9330': 0.5,
 '#9331': 0.5,
 '#9340': 0.020495,
 '#9341': 0.979505,
 '#9350': 0.62101,
 '#9351': 0.37899,
 '#9360': 0.07873,
 '#9361': 0.92127,
 '0G': 0.17671,
 '2Z': 0.061502,
 '@1': 18.227,
 '@10': 7.355e-05,
 '@100': 0.003556,
 '@101': 0.126125,
 '@102': 0.00769,
 '@103': 5.393e-05,
 '@104': 0.027789,
 '@105': 0.1419,
 '@106': 0.005081,
 '@107': 59.6785,
 '@108': 0.028835,
 '@109': 0.00048383,
 '@11': 0.000405,
 '@110': 0.01703,
 '@111': 0.026029,
 '@112': 0.0005408,
 '@113': 0.0002112,
 '@114': 0.0001255,
 '@115': 0.039845,
 '@116': 1.17e-05,
 '@117': 0.0013545,
 '@118': 0.015663,
 '@119': 0.009494,
 '@12': 4.494e-05,
 '@120': 0.016848,
 '@121': 0.0092,
 '@122': 0.003342,
 '@12

## Data management

Similar to coinbase.py, this script saves the data in a specified format. These functions handles duplicates and sorts by date. For hourly data, it aligns to the hour.

### HyperliquidDataManager base class

In [0]:
#| echo: false
#| output: asis
show_doc(save_hyperliquid_file)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L781){target="_blank" style="float:right; font-size:smaller"}

### save_hyperliquid_file

```python
def save_hyperliquid_file(
    df, folder_path, file_name, type:str='parquet'
):
```

*Save a pandas DataFrame to a file in either CSV or Parquet format.*

Args:
    df (pandas.DataFrame): The DataFrame to save
    folder_path (str): Directory path where the file will be saved
    file_name (str): Name of the file without extension
    type (str, optional): File format - either "csv" or "parquet". Defaults to "parquet"

The function saves the DataFrame to the specified path, handling the file extension automatically.
For CSV files, the index is not saved. For Parquet files, default Parquet settings are used.
Creates the folder if it doesn't exist.

In [0]:
#| echo: false
#| output: asis
show_doc(HyperliquidDataManager)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L807){target="_blank" style="float:right; font-size:smaller"}

### HyperliquidDataManager

```python
def HyperliquidDataManager(
    ticker:NoneType=None, data_dir:str='../data/hyperliquid', interval:str='1h', file_type:str='parquet',
    update:bool=False, save:bool=False, refresh_hours:int=24, info:NoneType=None, verbose:bool=True,
    data_type:NoneType=None
):
```

*Base class for managing Hyperliquid data files.*

Handles reading, updating, and saving data for different data types (perp, spot, funding).
Data is stored in organized folders by data type, with files named by token and interval.

Args:
    ticker (str or list, optional): Token symbol(s) to manage. If None, uses all available tokens.
    data_dir (str, optional): Base directory for data storage. Defaults to "../data/hyperliquid"
    interval (str, optional): Time interval for data ("1m", "5m", "15m", "1h", "4h", "1d"). 
        Defaults to "1h". Not used for funding data.
    file_type (str, optional): File format - "parquet" or "csv". Defaults to "parquet"
    update (bool, optional): If True, checks for and downloads new data. Defaults to False
    save (bool, optional): If True, saves updated data back to file. Defaults to False
    refresh_hours (int, optional): Hours of data to refresh when updating. Defaults to 24
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.
    verbose (bool, optional): If True, prints progress messages. Defaults to True

Attributes:
    data_type (str): Type of data managed by this instance ("perp", "spot", "funding")
    data (dict): Dictionary mapping tickers to their DataFrames

### HyperliquidPerpManager derived class

In [0]:
#| echo: false
#| output: asis
show_doc(HyperliquidPerpManager)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L1038){target="_blank" style="float:right; font-size:smaller"}

### HyperliquidPerpManager

```python
def HyperliquidPerpManager(
    ticker:NoneType=None, data_dir:str='../data/hyperliquid', interval:str='1h', file_type:str='parquet',
    update:bool=False, save:bool=False, refresh_hours:int=24, info:NoneType=None, verbose:bool=True
):
```

*Manager for Hyperliquid perpetual futures data.*

Handles reading, updating, and saving perpetual price data (OHLCV).
Data is stored in the 'perp' subfolder with files named as {ticker}_{interval}.{file_type}

Args:
    ticker (str or list, optional): Token symbol(s) to manage. If None, uses all available perp tokens.
    data_dir (str, optional): Base directory for data storage. Defaults to "../data/hyperliquid"
    interval (str, optional): Time interval for data ("1m", "5m", "15m", "1h", "4h", "1d"). 
        Defaults to "1h".
    file_type (str, optional): File format - "parquet" or "csv". Defaults to "parquet"
    update (bool, optional): If True, checks for and downloads new data. Defaults to False
    save (bool, optional): If True, saves updated data back to file. Defaults to False
    refresh_hours (int, optional): Hours of data to refresh when updating. Defaults to 24
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.
    verbose (bool, optional): If True, prints progress messages. Defaults to True

Examples:
    # Load existing perp data for ETH
    manager = HyperliquidPerpManager(ticker="ETH", interval="1h", info=info)
    eth_data = manager.data["ETH"]

    # Update and save data for multiple tokens
    manager = HyperliquidPerpManager(
        ticker=["ETH", "BTC", "SOL"],
        interval="4h",
        update=True,
        save=True,
        refresh_hours=48,
        info=info
    )

    # Load all available perp tokens
    manager = HyperliquidPerpManager(update=True, save=True, info=info)

#### Example

In [37]:
#| eval: false
# Load and update ETH perpetual data with 1h interval
manager = HyperliquidPerpManager(
    ticker="ETH",
    interval="1h",
    refresh_hours = 24,
    update=True,
    save=True,
    info=info
)
eth_data = manager.get_data("ETH")
print(eth_data)

Processing 1 perpetual tokens...
Loaded 26 rows for ETH from ../data/hyperliquid/perp/ETH_1h.parquet
Updating ETH from 2026-07-25T23:00:00Z
Retrieved 25 new rows for ETH
Added 0 new rows for ETH
Saved 26 rows for ETH
              datetime    open    high     low   close      volume coin
0  2026-07-25 22:00:00  1871.8  1876.0  1871.6  1875.2   2288.8254  ETH
1  2026-07-25 23:00:00  1875.2  1875.6  1872.6  1874.3   1635.3091  ETH
2  2026-07-26 00:00:00  1874.3  1880.5  1873.5  1880.0   5964.0359  ETH
3  2026-07-26 01:00:00  1880.1  1881.0  1877.4  1877.5   1205.5172  ETH
4  2026-07-26 02:00:00  1877.6  1880.5  1876.5  1880.2    925.1730  ETH
5  2026-07-26 03:00:00  1880.2  1883.4  1878.2  1881.0   2411.0634  ETH
6  2026-07-26 04:00:00  1881.4  1888.8  1879.6  1885.7   6799.9422  ETH
7  2026-07-26 05:00:00  1885.7  1886.4  1882.0  1882.3   3010.1127  ETH
8  2026-07-26 06:00:00  1882.4  1884.3  1877.6  1884.0   5390.1696  ETH
9  2026-07-26 07:00:00  1884.0  1885.5  1879.8  1880.5   1183.0

In [38]:
#|eval: false
manager = HyperliquidPerpManager(
    update=False,
    save=False,
    verbose=False
)
df = manager.get_data()
print(df)

                      datetime    open    high     low   close    volume coin
0    2025-03-19 13:00:00+00:00  1.3062  1.3101  1.2951  1.2974   1788.20  FTT
1    2025-03-19 14:00:00+00:00  1.2989  1.3067  1.2944  1.2976   1557.00  FTT
2    2025-03-19 15:00:00+00:00  1.2988  1.3105  1.2958  1.3088   2618.00  FTT
3    2025-03-19 16:00:00+00:00  1.3070  1.3134  1.3009  1.3026   1967.90  FTT
4    2025-03-19 17:00:00+00:00  1.3029  1.3029  1.2845  1.2895   1485.00  FTT
...                        ...     ...     ...     ...     ...       ...  ...
5669 2025-11-19 18:00:00+00:00  0.2272  0.2291  0.2224  0.2238  63048.07  ACE
5670 2025-11-19 19:00:00+00:00  0.2245  0.2275  0.2238  0.2265  57691.48  ACE
5671 2025-11-19 20:00:00+00:00  0.2264  0.2313  0.2261  0.2308  78869.60  ACE
5672 2025-11-19 21:00:00+00:00  0.2312  0.2373  0.2312  0.2371  44765.71  ACE
5673 2025-11-19 22:00:00+00:00  0.2369  0.2369  0.2345  0.2360  19050.64  ACE

[854220 rows x 7 columns]


### HyperliquidSpotManager derived class

In [0]:
#| echo: false
#| output: asis
show_doc(HyperliquidSpotManager)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L1317){target="_blank" style="float:right; font-size:smaller"}

### HyperliquidSpotManager

```python
def HyperliquidSpotManager(
    ticker:NoneType=None, base:str='USDC', data_dir:str='../data/hyperliquid', interval:str='1h',
    file_type:str='parquet', update:bool=False, save:bool=False, refresh_hours:int=24, info:NoneType=None,
    verbose:bool=True
):
```

*Manager for Hyperliquid spot market data.*

Handles reading, updating, and saving spot price data (OHLCV).
Data is stored in the 'spot' subfolder with files named as {ticker}_{base}_{interval}.{file_type}

Args:
    ticker (str or list, optional): Token symbol(s) to manage. If None, uses all available spot tokens.
    base (str, optional): Base currency for spot pairs. Defaults to "USDC".
    data_dir (str, optional): Base directory for data storage. Defaults to "../data/hyperliquid"
    interval (str, optional): Time interval for data ("1m", "5m", "15m", "1h", "4h", "1d"). 
        Defaults to "1h".
    file_type (str, optional): File format - "parquet" or "csv". Defaults to "parquet"
    update (bool, optional): If True, checks for and downloads new data. Defaults to False
    save (bool, optional): If True, saves updated data back to file. Defaults to False
    refresh_hours (int, optional): Hours of data to refresh when updating. Defaults to 24
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.
    verbose (bool, optional): If True, prints progress messages. Defaults to True

Examples:
    # Load existing spot data for ETH/USDC
    manager = HyperliquidSpotManager(ticker="ETH", base="USDC", interval="1h", info=info)
    eth_data = manager.data["ETH"]

    # Update and save data for multiple tokens
    manager = HyperliquidSpotManager(
        ticker=["ETH", "BTC", "SOL"],
        base="USDC",
        interval="4h",
        update=True,
        save=True,
        refresh_hours=48,
        info=info
    )

    # Load all available spot tokens
    manager = HyperliquidSpotManager(base="USDC", update=True, save=True, info=info)

#### Example

In [40]:
#| eval: false
# Example 1: Load existing spot data for a single token
print("Example 1: Load existing ETH/USDC spot data")
manager1 = HyperliquidSpotManager(
    ticker="ETH",
    base="USDC",
    interval="1h",
    update=True,
    save=True,
    info=info,
    verbose=True
)
if "ETH" in manager1.data:
    print(f"Loaded {len(manager1.data['ETH'])} records for ETH")
    print(manager1.data["ETH"])

Example 1: Load existing ETH/USDC spot data
Loaded 26 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_1h.parquet
Updating ETH/USDC from 2026-07-25T23:00:00Z
Retrieved 25 new rows for ETH/USDC
Added 0 new rows for ETH/USDC
Saved 26 rows for ETH/USDC
Loaded 26 records for ETH
              datetime    open    high     low   close    volume coin
0  2026-07-25 22:00:00  1872.5  1875.3  1872.4  1875.3   16.5510  ETH
1  2026-07-25 23:00:00  1874.4  1874.8  1872.2  1873.3   59.5098  ETH
2  2026-07-26 00:00:00  1873.5  1879.4  1872.9  1879.4   52.7933  ETH
3  2026-07-26 01:00:00  1880.1  1880.1  1876.5  1876.9   54.5726  ETH
4  2026-07-26 02:00:00  1877.3  1880.2  1876.0  1879.5   45.9965  ETH
5  2026-07-26 03:00:00  1879.5  1882.5  1877.8  1880.6   46.6177  ETH
6  2026-07-26 04:00:00  1881.1  1887.2  1879.1  1884.0  600.0430  ETH
7  2026-07-26 05:00:00  1884.7  1885.4  1881.9  1881.9   30.9801  ETH
8  2026-07-26 06:00:00  1881.8  1883.8  1877.3  1883.2  214.0866  ETH
9  2026-07-26 07:00:0

In [41]:
#| eval: false
# Example 2: Update and save data for multiple tokens
print("\nExample 2: Update and save multiple tokens")
manager2 = HyperliquidSpotManager(
    ticker=["ETH", "BTC", "SOL"],
    base="USDC",
    interval="1h",
    update=True,
    save=True,
    refresh_hours=24,
    info=info,
    verbose=True
)

# Access the data
for ticker in ["ETH", "BTC", "SOL"]:
    if ticker in manager2.data:
        print(f"\n{ticker} data shape: {manager2.data[ticker].shape}")
        print(f"Latest {ticker} price: ${manager2.data[ticker]['close'].iloc[-1]:.2f}")


Example 2: Update and save multiple tokens
Loaded 26 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_1h.parquet
Updating ETH/USDC from 2026-07-25T23:00:00Z
Retrieved 25 new rows for ETH/USDC
Added 0 new rows for ETH/USDC
Saved 26 rows for ETH/USDC
Loaded 26 rows for BTC from ../data/hyperliquid/spot/BTC_USDC_1h.parquet
Updating BTC/USDC from 2026-07-25T23:00:00Z
Retrieved 25 new rows for BTC/USDC
Added 0 new rows for BTC/USDC
Saved 26 rows for BTC/USDC
Loaded 26 rows for SOL from ../data/hyperliquid/spot/SOL_USDC_1h.parquet
Updating SOL/USDC from 2026-07-25T23:00:00Z
Retrieved 25 new rows for SOL/USDC
Added 0 new rows for SOL/USDC
Saved 26 rows for SOL/USDC

ETH data shape: (26, 7)
Latest ETH price: $1959.40

BTC data shape: (26, 7)
Latest BTC price: $65393.00

SOL data shape: (26, 7)
Latest SOL price: $76.90


In [42]:
#| eval: false
# Example 3: Load all available spot tokens with USDC base
print("\nExample 3: Load all available spot tokens")
manager3 = HyperliquidSpotManager(
    base="USDC",
    interval="4h",
    update=True,
    save=True,
    info=info,
    verbose=True
)

print(f"\nLoaded {len(manager3.data)} spot tokens")
print(f"Available tokens: {list(manager3.data.keys())}")


Example 3: Load all available spot tokens
Found 7 spot tokens for USDC
Loaded 7 rows for BTC from ../data/hyperliquid/spot/BTC_USDC_4h.parquet
Updating BTC/USDC from 2026-07-25T20:00:00Z
Retrieved 7 new rows for BTC/USDC
Added 0 new rows for BTC/USDC
Saved 7 rows for BTC/USDC
Loaded 7 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_4h.parquet
Updating ETH/USDC from 2026-07-25T20:00:00Z
Retrieved 7 new rows for ETH/USDC
Added 0 new rows for ETH/USDC
Saved 7 rows for ETH/USDC
Loaded 7 rows for SOL from ../data/hyperliquid/spot/SOL_USDC_4h.parquet
Updating SOL/USDC from 2026-07-25T20:00:00Z
Retrieved 7 new rows for SOL/USDC
Added 0 new rows for SOL/USDC
Saved 7 rows for SOL/USDC
Loaded 7 rows for HYPE from ../data/hyperliquid/spot/HYPE_USDC_4h.parquet
Updating HYPE/USDC from 2026-07-25T20:00:00Z
Retrieved 7 new rows for HYPE/USDC
Added 0 new rows for HYPE/USDC
Saved 7 rows for HYPE/USDC
Loaded 1621 rows for TRUMP from ../data/hyperliquid/spot/TRUMP_USDC_4h.parquet
Updating TRUMP/USDC

In [43]:
#| eval: false
# Example 5: Analyze spot data
print("\nExample 5: Analyze spot data")
manager5 = HyperliquidSpotManager(
    ticker=["ETH", "BTC"],
    base="USDC",
    interval="1h",
    update=True,
    info=info,
    verbose=True
)

# Calculate some basic statistics
for ticker in ["ETH", "BTC"]:
    if ticker in manager5.data:
        df = manager5.data[ticker]
        
        # Calculate 24h change
        if len(df) >= 24:
            price_24h_ago = df['close'].iloc[-24]
            current_price = df['close'].iloc[-1]
            change_24h = ((current_price - price_24h_ago) / price_24h_ago) * 100
            
            print(f"\n{ticker}/USDC:")
            print(f"  Current Price: ${current_price:.2f}")
            print(f"  24h Change: {change_24h:+.2f}%")
            print(f"  24h High: ${df['high'].iloc[-24:].max():.2f}")
            print(f"  24h Low: ${df['low'].iloc[-24:].min():.2f}")
            print(f"  24h Volume: {df['volume'].iloc[-24:].sum():.2f}")


Example 5: Analyze spot data
Loaded 26 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_1h.parquet
Updating ETH/USDC from 2026-07-25T23:00:00Z
Retrieved 25 new rows for ETH/USDC
Added 0 new rows for ETH/USDC
Loaded 26 rows for BTC from ../data/hyperliquid/spot/BTC_USDC_1h.parquet
Updating BTC/USDC from 2026-07-25T23:00:00Z
Retrieved 25 new rows for BTC/USDC
Added 0 new rows for BTC/USDC

ETH/USDC:
  Current Price: $1958.30
  24h Change: +4.20%
  24h High: $1964.70
  24h Low: $1872.90
  24h Volume: 4595.75

BTC/USDC:
  Current Price: $65346.00
  24h Change: +1.42%
  24h High: $65494.00
  24h Low: $64230.00
  24h Volume: 140.47


In [44]:
#|eval: false
mananger6 = HyperliquidSpotManager(
    base="USDC",
    interval="1h",
    update=False,
    save=False,
    verbose=True
)
mananger6.get_data()

Found 7 spot tokens for USDC
Loaded 26 rows for BTC from ../data/hyperliquid/spot/BTC_USDC_1h.parquet
Loaded 26 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_1h.parquet
Loaded 26 rows for SOL from ../data/hyperliquid/spot/SOL_USDC_1h.parquet
Loaded 484 rows for HYPE from ../data/hyperliquid/spot/HYPE_USDC_1h.parquet
Loaded 36 rows for TRUMP from ../data/hyperliquid/spot/TRUMP_USDC_1h.parquet
Loaded 25 rows for BERA from ../data/hyperliquid/spot/BERA_USDC_1h.parquet
Loaded 480 rows for PUMP from ../data/hyperliquid/spot/PUMP_USDC_1h.parquet


,datetime,open,high,low,close,volume,coin
0,2026-07-25 22:00:00+00:00,64303.000000,64356.000000,64290.000000,64350.000000,2.847586e+01,BTC
1,2026-07-25 23:00:00+00:00,64351.000000,64363.000000,64300.000000,64305.000000,1.306880e+00,BTC
2,2026-07-26 00:00:00+00:00,64304.000000,64433.000000,64300.000000,64433.000000,5.077590e+00,BTC
3,2026-07-26 01:00:00+00:00,64437.000000,64481.000000,64399.000000,64402.000000,3.229600e+00,BTC
4,2026-07-26 02:00:00+00:00,64407.000000,64470.000000,64373.000000,64468.000000,1.359620e+00,BTC
...,...,...,...,...,...,...,...
475,2025-11-19 13:00:00+00:00,0.000148,0.000148,0.000148,0.000148,0.000000e+00,PUMP
476,2025-11-19 14:00:00+00:00,0.000148,0.000151,0.000148,0.000151,1.585506e+06,PUMP
477,2025-11-19 15:00:00+00:00,0.000151,0.000151,0.000151,0.000151,0.000000e+00,PUMP
478,2025-11-19 16:00:00+00:00,0.000151,0.000151,0.000151,0.000151,0.000000e+00,PUMP


### HyperliquidFundingManager derived class

In [0]:
#| echo: false
#| output: asis
show_doc(HyperliquidFundingManager)

---

[source](https://github.com/silvaac/token_data/blob/main/token_data/hyperliquid.py#L1584){target="_blank" style="float:right; font-size:smaller"}

### HyperliquidFundingManager

```python
def HyperliquidFundingManager(
    ticker:NoneType=None, data_dir:str='../data/hyperliquid', file_type:str='parquet', update:bool=False,
    save:bool=False, refresh_hours:int=24, round_to_hour:bool=True, info:NoneType=None, verbose:bool=True
):
```

*Manager for Hyperliquid funding rate data.*

Handles reading, updating, and saving funding rate data.
Data is stored in the 'funding' subfolder with files named as {ticker}.{file_type}

Args:
    ticker (str or list, optional): Token symbol(s) to manage. If None, uses all available tokens.
    data_dir (str, optional): Base directory for data storage. Defaults to "../data/hyperliquid"
    file_type (str, optional): File format - "parquet" or "csv". Defaults to "parquet"
    update (bool, optional): If True, checks for and downloads new data. Defaults to False
    save (bool, optional): If True, saves updated data back to file. Defaults to False
    refresh_hours (int, optional): Hours of data to refresh when updating. Defaults to 24
    round_to_hour (bool, optional): If True, rounds datetime to nearest hour. Defaults to True
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.
    verbose (bool, optional): If True, prints progress messages. Defaults to True

Examples:
    # Load existing funding data for ETH
    manager = HyperliquidFundingManager(ticker="ETH", info=info)
    eth_data = manager.data["ETH"]

    # Update and save data for multiple tokens
    manager = HyperliquidFundingManager(
        ticker=["ETH", "BTC", "SOL"],
        update=True,
        save=True,
        refresh_hours=48,
        info=info
    )

    # Load all available tokens
    manager = HyperliquidFundingManager(update=True, save=True, info=info)

#### Example

In [46]:
#|eval: false
manager = HyperliquidFundingManager(ticker="ETH")
eth_data = manager.data["ETH"]
print(eth_data)

Loaded 738 rows for ETH from ../data/hyperliquid/funding/ETH.parquet
                     datetime  funding_rate   premium coin  fund_calc
0   2025-10-03 21:00:00+00:00      0.000013  0.000185  ETH   0.000013
1   2025-10-03 22:00:00+00:00      0.000013  0.000111  ETH   0.000013
2   2025-10-03 23:00:00+00:00      0.000013  0.000141  ETH   0.000013
3   2025-10-04 00:00:00+00:00      0.000013  0.000113  ETH   0.000013
4   2025-10-04 01:00:00+00:00      0.000013  0.000081  ETH   0.000013
..                        ...           ...       ...  ...        ...
733 2025-11-19 07:00:00+00:00      0.000013  0.000175  ETH   0.000013
734 2025-11-19 08:00:00+00:00      0.000013  0.000281  ETH   0.000013
735 2025-11-19 09:00:00+00:00      0.000013  0.000277  ETH   0.000013
736 2025-11-19 10:00:00+00:00      0.000013  0.000256  ETH   0.000013
737 2025-11-19 11:00:00+00:00      0.000013  0.000295  ETH   0.000013

[738 rows x 5 columns]


In [47]:
#|eval: false
manager = HyperliquidFundingManager(update=False, save=False,verbose=False)
manager.get_data()

,datetime,funding_rate,premium,coin,fund_calc
0,2025-10-03 21:00:00+00:00,-0.000013,-0.000606,FTT,-0.000013
1,2025-10-03 22:00:00+00:00,0.000013,-0.000313,FTT,0.000013
2,2025-10-03 23:00:00+00:00,0.000013,-0.000128,FTT,0.000013
3,2025-10-04 00:00:00+00:00,0.000013,-0.000041,FTT,0.000013
4,2025-10-04 01:00:00+00:00,0.000013,-0.000069,FTT,0.000013
...,...,...,...,...,...
733,2025-11-19 07:00:00+00:00,0.000013,-0.000111,ACE,0.000013
734,2025-11-19 08:00:00+00:00,0.000013,-0.000264,ACE,0.000013
735,2025-11-19 09:00:00+00:00,0.000013,-0.000378,ACE,0.000013
736,2025-11-19 10:00:00+00:00,0.000013,-0.000113,ACE,0.000013


## More examples

Load package:

In [48]:
#|eval: false
from token_data.hyperliquid import *

Load all available funding rates already stored in the data directory. It returns a pandas DataFrame with columns 'datetime' in UTC and 'funding_rate' as the funding rate used within the Hyperliquid system to find the USDC paid or received for a given perpetual position.

The other coluns are "premium" and "fund_calc". The "premium" column represents the premium used in the funding rate calculation and the "fund_calc" is currently very experimental and tries to replicated the funding rate calculation in the Hyperliquid system. Use "funding_rate" column in simulations or P&L calculations.

In [49]:
#|eval: false
manager = HyperliquidFundingManager(update=False, save=False, verbose=False)
manager.get_data()

,datetime,funding_rate,premium,coin,fund_calc
0,2025-10-03 21:00:00+00:00,-0.000013,-0.000606,FTT,-0.000013
1,2025-10-03 22:00:00+00:00,0.000013,-0.000313,FTT,0.000013
2,2025-10-03 23:00:00+00:00,0.000013,-0.000128,FTT,0.000013
3,2025-10-04 00:00:00+00:00,0.000013,-0.000041,FTT,0.000013
4,2025-10-04 01:00:00+00:00,0.000013,-0.000069,FTT,0.000013
...,...,...,...,...,...
733,2025-11-19 07:00:00+00:00,0.000013,-0.000111,ACE,0.000013
734,2025-11-19 08:00:00+00:00,0.000013,-0.000264,ACE,0.000013
735,2025-11-19 09:00:00+00:00,0.000013,-0.000378,ACE,0.000013
736,2025-11-19 10:00:00+00:00,0.000013,-0.000113,ACE,0.000013


Load all available spot prices already stored in the data directory. It returns a pandas DataFrame with columns 'datetime' in UTC and 'open', 'high', 'low', 'close', 'volume' and 'coin'. Important to note that the list of spots is limited and was hard coded. In the future this can be changed, therefore, the list is limited to some of the most important tokens.

In [50]:
#|eval: false
manager = HyperliquidSpotManager(update=False, save=False, verbose=False)
manager.get_data()

,datetime,open,high,low,close,volume,coin
0,2026-07-25 22:00:00+00:00,64303.000000,64356.000000,64290.000000,64350.000000,2.847586e+01,BTC
1,2026-07-25 23:00:00+00:00,64351.000000,64363.000000,64300.000000,64305.000000,1.306880e+00,BTC
2,2026-07-26 00:00:00+00:00,64304.000000,64433.000000,64300.000000,64433.000000,5.077590e+00,BTC
3,2026-07-26 01:00:00+00:00,64437.000000,64481.000000,64399.000000,64402.000000,3.229600e+00,BTC
4,2026-07-26 02:00:00+00:00,64407.000000,64470.000000,64373.000000,64468.000000,1.359620e+00,BTC
...,...,...,...,...,...,...,...
475,2025-11-19 13:00:00+00:00,0.000148,0.000148,0.000148,0.000148,0.000000e+00,PUMP
476,2025-11-19 14:00:00+00:00,0.000148,0.000151,0.000148,0.000151,1.585506e+06,PUMP
477,2025-11-19 15:00:00+00:00,0.000151,0.000151,0.000151,0.000151,0.000000e+00,PUMP
478,2025-11-19 16:00:00+00:00,0.000151,0.000151,0.000151,0.000151,0.000000e+00,PUMP


Next example loads perpetual prices already stored in the data directory on an hourly frequency (default but can be changed). It returns a pandas DataFrame with columns 'datetime' in UTC and 'open', 'high', 'low', 'close', 'volume' and 'coin'.

In [51]:
#|eval: false
manager = HyperliquidPerpManager(update=False, save=False,verbose=False)
manager.get_data()

,datetime,open,high,low,close,volume,coin
0,2025-03-19 13:00:00+00:00,1.3062,1.3101,1.2951,1.2974,1788.20,FTT
1,2025-03-19 14:00:00+00:00,1.2989,1.3067,1.2944,1.2976,1557.00,FTT
2,2025-03-19 15:00:00+00:00,1.2988,1.3105,1.2958,1.3088,2618.00,FTT
3,2025-03-19 16:00:00+00:00,1.3070,1.3134,1.3009,1.3026,1967.90,FTT
4,2025-03-19 17:00:00+00:00,1.3029,1.3029,1.2845,1.2895,1485.00,FTT
...,...,...,...,...,...,...,...
5669,2025-11-19 18:00:00+00:00,0.2272,0.2291,0.2224,0.2238,63048.07,ACE
5670,2025-11-19 19:00:00+00:00,0.2245,0.2275,0.2238,0.2265,57691.48,ACE
5671,2025-11-19 20:00:00+00:00,0.2264,0.2313,0.2261,0.2308,78869.60,ACE
5672,2025-11-19 21:00:00+00:00,0.2312,0.2373,0.2312,0.2371,44765.71,ACE


Next example loads only "ETH" perpetual prices already stored in the data directory on an hourly frequency (default but can be changed). It returns a pandas DataFrame with columns 'datetime' in UTC.

In [52]:
#|eval: false
manager = HyperliquidPerpManager(ticker='ETH',update=False, save=False,verbose=True)
manager.get_data()

Processing 1 perpetual tokens...
Loaded 26 rows for ETH from ../data/hyperliquid/perp/ETH_1h.parquet


,datetime,open,high,low,close,volume,coin
0,2026-07-25 22:00:00+00:00,1871.8,1876.0,1871.6,1875.2,2288.8254,ETH
1,2026-07-25 23:00:00+00:00,1875.2,1875.6,1872.6,1874.3,1635.3091,ETH
2,2026-07-26 00:00:00+00:00,1874.3,1880.5,1873.5,1880.0,5964.0359,ETH
3,2026-07-26 01:00:00+00:00,1880.1,1881.0,1877.4,1877.5,1205.5172,ETH
4,2026-07-26 02:00:00+00:00,1877.6,1880.5,1876.5,1880.2,925.1730,ETH
5,2026-07-26 03:00:00+00:00,1880.2,1883.4,1878.2,1881.0,2411.0634,ETH
6,2026-07-26 04:00:00+00:00,1881.4,1888.8,1879.6,1885.7,6799.9422,ETH
7,2026-07-26 05:00:00+00:00,1885.7,1886.4,1882.0,1882.3,3010.1127,ETH
8,2026-07-26 06:00:00+00:00,1882.4,1884.3,1877.6,1884.0,5390.1696,ETH
9,2026-07-26 07:00:00+00:00,1884.0,1885.5,1879.8,1880.5,1183.0028,ETH


If you want to make sure you are featching the most recent data, use the `update=True` parameter. In that case one needs to pass the `info` parameter with your API key to retrieve real-time data which is setup beforehand. The next example demonstrates how to use this:

In [53]:
#|eval: false
address, info, exchange = setup(base_url=constants.MAINNET_API_URL, skip_ws=True)
manager = HyperliquidPerpManager(ticker='ETH',update=True, save=False, verbose=True,info=info)
manager.get_data()

Processing 1 perpetual tokens...
Loaded 26 rows for ETH from ../data/hyperliquid/perp/ETH_1h.parquet
Updating ETH from 2026-07-25T23:00:00Z
Retrieved 25 new rows for ETH
Added 0 new rows for ETH


,datetime,open,high,low,close,volume,coin
0,2026-07-25 22:00:00,1871.8,1876.0,1871.6,1875.2,2288.8254,ETH
1,2026-07-25 23:00:00,1875.2,1875.6,1872.6,1874.3,1635.3091,ETH
2,2026-07-26 00:00:00,1874.3,1880.5,1873.5,1880.0,5964.0359,ETH
3,2026-07-26 01:00:00,1880.1,1881.0,1877.4,1877.5,1205.5172,ETH
4,2026-07-26 02:00:00,1877.6,1880.5,1876.5,1880.2,925.1730,ETH
5,2026-07-26 03:00:00,1880.2,1883.4,1878.2,1881.0,2411.0634,ETH
6,2026-07-26 04:00:00,1881.4,1888.8,1879.6,1885.7,6799.9422,ETH
7,2026-07-26 05:00:00,1885.7,1886.4,1882.0,1882.3,3010.1127,ETH
8,2026-07-26 06:00:00,1882.4,1884.3,1877.6,1884.0,5390.1696,ETH
9,2026-07-26 07:00:00,1884.0,1885.5,1879.8,1880.5,1183.0028,ETH


The same goes for the other data items. Such as spot prices and funding rates. Next example shows the spot prices for "ETH" on an hourly frequency and updates it:

In [54]:
#|eval: false
manager = HyperliquidSpotManager(ticker='ETH',update=True, save=False, verbose=True,info=info)
manager.get_data()

Loaded 26 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_1h.parquet
Updating ETH/USDC from 2026-07-25T23:00:00Z
Retrieved 25 new rows for ETH/USDC
Added 0 new rows for ETH/USDC


,datetime,open,high,low,close,volume,coin
0,2026-07-25 22:00:00,1872.5,1875.3,1872.4,1875.3,16.5510,ETH
1,2026-07-25 23:00:00,1874.4,1874.8,1872.2,1873.3,59.5098,ETH
2,2026-07-26 00:00:00,1873.5,1879.4,1872.9,1879.4,52.7933,ETH
3,2026-07-26 01:00:00,1880.1,1880.1,1876.5,1876.9,54.5726,ETH
4,2026-07-26 02:00:00,1877.3,1880.2,1876.0,1879.5,45.9965,ETH
5,2026-07-26 03:00:00,1879.5,1882.5,1877.8,1880.6,46.6177,ETH
6,2026-07-26 04:00:00,1881.1,1887.2,1879.1,1884.0,600.0430,ETH
7,2026-07-26 05:00:00,1884.7,1885.4,1881.9,1881.9,30.9801,ETH
8,2026-07-26 06:00:00,1881.8,1883.8,1877.3,1883.2,214.0866,ETH
9,2026-07-26 07:00:00,1883.9,1884.7,1879.2,1879.9,200.7441,ETH


If you want to update the funding rate, do the following:

In [55]:
#|eval: false
manager = HyperliquidFundingManager(ticker='ETH',update=True, save=False, verbose=True,info=info)
manager.get_data()

Loaded 738 rows for ETH from ../data/hyperliquid/funding/ETH.parquet
Updating ETH from 2025-11-18T11:00:00Z
Retrieved 500 new rows for ETH
Added 475 new rows for ETH


,datetime,funding_rate,premium,coin,fund_calc
0,2025-10-03 21:00:00,0.000013,0.000185,ETH,0.000013
1,2025-10-03 22:00:00,0.000013,0.000111,ETH,0.000013
2,2025-10-03 23:00:00,0.000013,0.000141,ETH,0.000013
3,2025-10-04 00:00:00,0.000013,0.000113,ETH,0.000013
4,2025-10-04 01:00:00,0.000013,0.000081,ETH,0.000013
...,...,...,...,...,...
1208,2025-12-09 02:00:00,0.000013,0.000120,ETH,0.000013
1209,2025-12-09 03:00:00,0.000013,0.000032,ETH,0.000013
1210,2025-12-09 04:00:00,0.000013,0.000102,ETH,0.000013
1211,2025-12-09 05:00:00,0.000013,0.000196,ETH,0.000013


## RWA / HIP-3 examples
The cells below demonstrate the auto-resolution of RWA tickers (e.g. `XLE` → `xyz:XLE`).

In [57]:
#| eval: false
# Resolve ticker
print("resolve_hyperliquid_ticker('ETH') ->", resolve_hyperliquid_ticker("ETH"))
print("resolve_hyperliquid_ticker('XLE') ->", resolve_hyperliquid_ticker("XLE"))
print("resolve_hyperliquid_ticker('xyz:XLE') ->", resolve_hyperliquid_ticker("xyz:XLE"))

resolve_hyperliquid_ticker('ETH') -> ETH
resolve_hyperliquid_ticker('XLE') -> xyz:XLE
resolve_hyperliquid_ticker('xyz:XLE') -> xyz:XLE


In [58]:
#| eval: false
# RWA perp price
xle_perp = retrieve_hyperliquid_perp_price(coin="XLE", interval="1h", info=info)
print("XLE perp price:")
print(xle_perp.head())
print(f"Columns: {list(xle_perp.columns)}")
print(f"coin value: {xle_perp['coin'].iloc[0]}")

XLE perp price:
             datetime    open    high     low   close  volume coin
0 2026-07-24 23:00:00  59.789  59.793  59.789  59.793    3.59  XLE
1 2026-07-25 00:00:00  59.960  59.960  59.559  59.642   46.08  XLE
2 2026-07-25 01:00:00  59.601  59.712  59.453  59.671  334.74  XLE
3 2026-07-25 02:00:00  59.742  59.835  59.692  59.834   68.50  XLE
4 2026-07-25 03:00:00  59.859  59.859  59.859  59.859    1.67  XLE
Columns: ['datetime', 'open', 'high', 'low', 'close', 'volume', 'coin']
coin value: XLE


In [59]:
#| eval: false
# RWA funding history
xle_funding = retrieve_hyperliquid_funding_history(coin="XLE", info=info)
print("XLE funding history:")
print(xle_funding.head())

XLE funding history:
                 datetime  funding_rate   premium coin  fund_calc
0 2026-07-25 00:00:00.066      0.000006  0.000206  XLE   0.000006
1 2026-07-25 01:00:00.036      0.000006 -0.000030  XLE   0.000006
2 2026-07-25 02:00:00.050     -0.000025 -0.000703  XLE  -0.000025
3 2026-07-25 03:00:00.016      0.000027  0.000731  XLE   0.000027
4 2026-07-25 04:00:00.036      0.000058  0.001233  XLE   0.000058


In [60]:
#| eval: false
# RWA mid price
xle_mid = hyperliquid_mids(coin="XLE", info=info)
print(f"XLE mid price: {xle_mid}")

XLE mid price: 58.78


In [61]:
#| eval: false
# RWA L2 snapshot
xle_l2 = retrieve_hyperliquid_l2_snapshot(coin="XLE", info=info)
print("XLE L2 snapshot:")
print(xle_l2.head())

XLE L2 snapshot:
                 datetime side   price    size  num_orders
0 2026-07-26 23:17:56.011  bid  58.735   17.03           1
1 2026-07-26 23:17:56.011  bid  58.734   54.97           1
2 2026-07-26 23:17:56.011  bid  58.733   17.03           1
3 2026-07-26 23:17:56.011  bid  58.732  855.76           2
4 2026-07-26 23:17:56.011  bid  58.731   39.28           1


In [62]:
#| eval: false
# List HIP-3 dex tokens
rwa_tokens = hyperliquid_tokens(info=info, dex="xyz")
print("xyz dex tokens:")
print(rwa_tokens.head(10))

xyz dex tokens:
    szDecimals        name  maxLeverage  marginTableId growthMode  \
0            4  xyz:XYZ100           30             30    enabled   
1            3    xyz:TSLA           20             20    enabled   
2            3    xyz:NVDA           20             20    enabled   
3            4    xyz:GOLD           25             25        NaN   
8            3    xyz:META           20             20    enabled   
9            3    xyz:AAPL           20             20    enabled   
10           3    xyz:MSFT           20             20    enabled   
12           3   xyz:GOOGL           20             20    enabled   
13           3    xyz:AMZN           20             20    enabled   
15           3      xyz:MU           10             10    enabled   

         lastGrowthModeChangeTime  onlyIsolated marginMode  isDelisted  
0   2025-11-23T17:37:10.033211662         False        NaN       False  
1   2025-11-23T17:37:10.033211662         False        NaN       False  
2   2

In [63]:
#| eval: false
# Regression check: ETH still works
eth_perp = retrieve_hyperliquid_perp_price(coin="ETH", interval="1h", info=info)
eth_funding = retrieve_hyperliquid_funding_history(coin="ETH", info=info)
print(f"ETH perp rows: {len(eth_perp)}, columns: {list(eth_perp.columns)}")
print(f"ETH funding rows: {len(eth_funding)}, columns: {list(eth_funding.columns)}")
print("ETH perp head:")
print(eth_perp.head())

ETH perp rows: 49, columns: ['datetime', 'open', 'high', 'low', 'close', 'volume', 'coin']
ETH funding rows: 48, columns: ['datetime', 'funding_rate', 'premium', 'coin', 'fund_calc']
ETH perp head:
             datetime    open    high     low   close     volume coin
0 2026-07-24 23:00:00  1858.8  1861.1  1858.1  1861.0  2792.8168  ETH
1 2026-07-25 00:00:00  1860.7  1863.7  1856.3  1859.3  4137.4891  ETH
2 2026-07-25 01:00:00  1859.3  1862.6  1857.4  1859.3  4557.0876  ETH
3 2026-07-25 02:00:00  1859.3  1860.7  1857.3  1858.3  4404.2636  ETH
4 2026-07-25 03:00:00  1858.3  1859.5  1854.8  1858.3  4298.4131  ETH
